airflow/plugins/operators/blackcat_to_gcs.py

```diff
@@ -0,0 +1,193 @@
+from calitp_data_infra.storage import get_fs, make_name_bq_safe
+from airflow.models import BaseOperator
+from pydantic import BaseModel
+from typing import Optional
+import pandas as pd
+import pendulum
+import requests
+import logging
+import gzip
+import os
+import re
+
+def write_to_log(logfilename):
+    '''
+    Creates a logger object that outputs to a log file, to the filename specified,
+    and also streams to console.
+    '''
+    logger = logging.getLogger(__name__)
+    logger.setLevel(logging.INFO)
+    formatter = logging.Formatter(f'%(asctime)s:%(levelname)s: %(message)s',
+                                  datefmt='%y-%m-%d %H:%M:%S')
+    file_handler = logging.FileHandler(logfilename)
+    file_handler.setFormatter(formatter)
+    stream_handler = logging.StreamHandler()
+    stream_handler.setFormatter(formatter)
+
+    if not logger.hasHandlers():
+        logger.addHandler(file_handler)
+        logger.addHandler(stream_handler)
+
+    return logger
+
+
+def camel_to_snake(name):
+    '''Converts Snake case to underscore separation for renaming columns; 
+    VehicleStatus becomes vehicle_status and 
+    can handle acroynms like ADAAccess, which becomes ada_access'''
+    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
+    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', name).lower()
+
+
+class BlackCatApiExtract(BaseModel):
+    api_url: str
+    form: str
+    api_tablename: str
+    bq_table_name: str
+    data: Optional[pd.DataFrame]
+    logger: Optional[logging.Logger]
+    extract_time: Optional[pendulum.DateTime]
+
+    logger = write_to_log('load_bc_apidata_output.log')
+    extract_time = pendulum.now()
+    
+    # pydantic doesn't know dataframe type
+    # see https://stackoverflow.com/a/69200069
+    class Config:
+        arbitrary_types_allowed = True
+    
+    def fetch_from_bc_api(self):
+        """Download a BlackCat table as a DataFrame.
+
+        Note that BlackCat API reports have rows structured as follows:
+        [{'ReportId': <id>,
+        'Organization': <organization>,
+        'ReportPeriod': <year>,
+        'ReportStatus': <status>,
+        'ReportLastModifiedDate': <timestamp>,
+        '<table_name>': {'Data': [{colname: value, ...}, {colname: value, ...} ...]}},
+        {'ReportId': <id>, ...etc. to the next organization}]
+
+        This function applies renames in the following order.
+            1. rename column names from snakecase to names utilizing underscores
+            2. rename fields
+            3. apply column prefix (to columns not renamed by 1 or 2)
+        """
+        
+        self.logger.info(
+            f"Downloading BlackCat data for {self.extract_time.format('YYYY')}_{self.bq_table_name}."
+        )
+        response = requests.get(self.api_url, verify=False)
+        blob = response.json()
+
+        org_data = []
+
+        # Cyling through and processing each org's data
+        for x in blob:
```

Comment:
This seems like it deviates from some other patterns we have when downloading raw data. In some other DAGs we save the raw JSON or GTFS files to cloud storage before extracting anything from them. After saving the raw contents, there can be further operations that happen on top of that. If only a part of the JSON is saved after getting a response from the server, any other elements in the JSON file are lost forever. 

In a much earlier version of the pipeline we downloaded GTFS data and extracted the files instead of saving the entire zip file. It came back to bite us when we wanted to examine the entire zip file.

Reply:
changed as suggested!